### PR TITLE
Bump to v6.6.4 and apply Pegasus changes

### PR DESCRIPTION
**Pull request by acelyc111**
## Changes
- Bump to v6.6.4.
  Since v5.9.2, there are some performance improvements and bugfixes, and also some new features like `atomic_flush` we want to use.
- Apply Pegasus changes.

## Tips

- To review this PR friendly, you could compare this PR with https://github.com/XiaoMi/pegasus-rocksdb/pull/23, which is the version we currently use and contains all the changes we make on rocksdb v5.9.2.

- The failed CI cases should be ignored, cause they are not related with this PR changes, and they are already failed on offical branch [CI](https://travis-ci.org/facebook/rocksdb/builds/644543782?utm_source=github_status&utm_medium=notification)

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -121,4 +121,4 @@ script:
     esac
```
Comment (hycdong) [inline .travis.yml:121]: Compared with #23, there are some difference in `TEST_GROUP`. 
For instance, in 5.9.2, one of them is below: 
```
if [ "${TEST_GROUP}" == '1' ]; then OPT=-DTRAVIS V=1 ROCKSDBTESTS_START=db_block_cache_test ROCKSDBTESTS_END=comparator_db_test make -j4 check_some; fi
```
in pegasus version, `ROCKSDBTESTS_END` changed.
```
if [ "${TEST_GROUP}" == '1' ]; then OPT=-DTRAVIS V=1 ROCKSDBTESTS_START=db_block_cache_test  ROCKSDBTESTS_END=db_log_iter_test make -j4 check_some; fi
```
I wonder that ci will be passed if we apply those test group changes.

File: db/db_filesnapshot.cc [hunk 2]
```diff
@@ -147,6 +147,63 @@ Status DBImpl::GetLiveFiles(std::vector<std::string>& ret,
   return Status::OK();
```
Comment (hycdong) [inline db/db_filesnapshot.cc:147]: function `EnableFileDeletions` changes a lot, pegasus version add some checks and log printing in if branch `if (disable_delete_obsolete_files_ == 0)`. Is this useless in v6.6.4?

File: db/db_impl/db_impl_compaction_flush.cc [hunk 3]
```diff
@@ -1605,6 +1641,7 @@ Status DBImpl::FlushMemTable(ColumnFamilyData* cfd,
         nonmem_write_thread_.ExitUnbatched(&nonmem_w);
       }
     }
+    }
```
Comment (hycdong) [inline db/db_impl/db_impl_compaction_flush.cc]: Format, not adding indent here.

File: db/version_edit.cc [hunk 4]
```diff
@@ -36,7 +36,8 @@ enum Tag : uint32_t {
   kNewFile = 7,
   // 8 was used for large value refs
   kPrevLogNumber = 9,
-  kMinLogNumberToKeep = 10,
+  kLastFlushSeqDecree = 10,
+  kMinLogNumberToKeep = 11,
```
Comment (hycdong) [inline db/version_edit.cc:44]: I suggest adding comment here to explain difference between pegasus version and official 6.6.4

File: db/version_set.cc [hunk 5]
```diff
@@ -3423,6 +3425,7 @@ std::string Version::DebugString(bool hex, bool print_stats) const {
       AppendNumberTo(&r, files[i]->fd.largest_seqno);
       r.append("]");
       r.append("[");
+      // TODO(laiyingchun): Pegasus data DebugString() not work correctly now.
```
Comment (hycdong) [inline db/version_set.cc]: Could you please add more comments to explain why function DebugString() not work correctly for pegasus data?

File: db/version_set.cc [hunk 6]
```diff
@@ -4254,6 +4293,21 @@ Status VersionSet::ExtractInfoFromVersionEdit(
...
+  if (from_edit.has_pegasus_data_version_) {
+    version_edit_params->SetPegasusDataVersion(from_edit.pegasus_data_version_);
+  }
+  
+  if (from_edit.has_last_manual_compact_finish_time_) {
+    version_edit_params->SetLastManualCompactFinishTime(
+        from_edit.last_manual_compact_finish_time_);
+  }
+  
+  if (from_edit.has_last_flush_seq_decree_) {
```
Comment (hycdong) [inline db/version_set.cc:4305]: Do we need check if current edit's last_flush_sequence is greater than from_edit just like the code below? or just update those value directly?
```
auto cur_last_flush_seq = version_edit_params->get_last_flush_seq();
auto cur_last_flush_decree = version_edit_params->get_last_flush_decree();

if(cur_last_flush_seq > from_edit.last_flush_sequence_ && 
   cur_last_flush_decree > from_edit.last_flush_decree_){
    version_edit_params->UpdateLastFlushSeqDecree(
        from_edit.last_flush_sequence_, from_edit.last_flush_decree_);
}
```

File: .travis.yml (hunk 1 above)
Comment (acelyc111) [reply .travis.yml:121]: Yes, we didn't apply these changes and some CI failed.
I think it's OK, we should make changes as less as possible, and we are rear to modify rocksdb, manual check the failed CI is accptable.

File: db/db_filesnapshot.cc (hunk 2 above)
Comment (acelyc111) [reply db/db_filesnapshot.cc:147]: Yes, there is a bug in v5.9.2, and v6.6.4 has fixed it by commit 46e599fc6b and 9da88a8321.
The log we added it useless, I removed it.

File: .travis.yml (hunk 1 above)
Comment (hycdong) [reply .travis.yml:121]: Okay~

File: db/db_impl/db_impl_compaction_flush.cc (hunk 3 above)
Comment (acelyc111) [reply db/db_impl/db_impl_compaction_flush.cc]: Yes, I didn't add indent here intentionally to avoid change the whole block.

File: db/version_set.cc (hunk 5 above)
Comment (acelyc111) [reply db/version_set.cc]: Formerly I considered that keys are encoded in Pegasus format, DebugString the keys is not readable, but now I think this TODO comment is redundant, because keys could be print in hex mode. I'll remove it.

File: db/version_set.cc (hunk 6 above)
Comment (acelyc111) [reply db/version_set.cc:4305]: You're right! This judgement has been implemented in UpdateLastFlushSeqDecree.

Comment (hycdong) [reply db/version_set.cc:4305]: Oh, I find it in `version_edit.h`

**Review by hycdong — APPROVED** [review]

File: include/rocksdb/db.h [hunk 7]
```diff
@@ -1168,6 +1176,14 @@ class DB {
                               uint64_t* manifest_file_size,
                               bool flush_memtable = true) = 0;
 
+  // Retrieve the list of all files in the database.
+  virtual Status GetLiveFilesQuick(std::vector<std::string>& /*ret*/,
+                                   uint64_t* /*manifest_file_size*/,
+                                   SequenceNumber* /*last_sequence*/,
+                                   uint64_t* /*last_decree*/) const {
```
Comment (levy5307) [inline include/rocksdb/db.h:1183]: I think we can format the code here

File: tools/sst_dump_tool.cc [hunk 8]
```diff
@@ -381,6 +493,21 @@ Status SstFileDumper::ReadSequential(bool print_kv, uint64_t read_num,
...
+        if (ikey.user_key.size() >= 2) {
+          uint32_t expire_ts = 0;
+          std::string hash_key, sort_key, user_data;
+          pegasus_restore_key(ikey.user_key, hash_key, sort_key);
+          pegasus_restore_value(value, expire_ts, user_data);
+          std::ostringstream oss;
+          oss << "\"" << escape_string(hash_key) << "\" : \"" << escape_string(sort_key)
+              << "\" @ " << ikey.sequence << " : " << ikey.type << " => "
+              << expire_ts << " : \"" << escape_string(user_data) << "\"";
+          fprintf(stdout, "%s\n", oss.str().c_str());
```
Comment (levy5307) [inline tools/sst_dump_tool.cc:506]: In pr 23, If it is not pegasus_data should also print some info. Is it useless?
code in pr 23:
```
 if (!options_.pegasus_data) {
        fprintf(stdout, "%s => %s\n",
            ikey.DebugString(output_hex_).c_str(),
            value.ToString(output_hex_).c_str());
      } else {
        if (ikey.user_key.size() >= 2) {
          uint32_t expire_ts = 0;
          std::string hash_key, sort_key, user_data;
          pegasus_restore_key(ikey.user_key, hash_key, sort_key);
          pegasus_restore_value(value, expire_ts, user_data);
          std::ostringstream oss;
          oss << "\"" << escape_string(hash_key) << "\" : \"" << escape_string(sort_key)
              << "\" @ " << ikey.sequence << " : " << ikey.type << " => "
              << expire_ts << " : \"" << escape_string(user_data) << "\"";
          fprintf(stdout, "%s\n", oss.str().c_str());
        }
      }
```

File: db/db_impl/db_impl_compaction_flush.cc (hunk 3 above)
Comment (neverchanje) [reply db/db_impl/db_impl_compaction_flush.cc]: You can add a comment here.

```suggestion
    } // if (!pegasus_data_ || !cfd->mem()->IsEmpty())
```

File: db/version_edit.cc [hunk 9]
```diff
@@ -48,6 +53,8 @@ enum Tag : uint32_t {
   kColumnFamilyAdd = 201,
   kColumnFamilyDrop = 202,
   kMaxColumnFamily = 203,
+  kValueSchemaVersion = 204,
```
Comment (neverchanje) [inline db/version_edit.cc]: call it `kPegasusDataVersion`? Because there's no such thing as "value schema version" now.

File: util/compression.h [hunk 10]
```diff
@@ -1048,7 +1048,6 @@ inline bool LZ4_Compress(const CompressionInfo& info,
 #else   // up to r123
   outlen = LZ4_compress_limitedOutput(input, &(*output)[output_header_len],
                                       static_cast<int>(length), compress_bound);
-  (void)ctx;
```
Comment (neverchanje) [inline util/compression.h:1051]: is this change necessary?

File: db/write_batch.cc [hunk 11]
```diff
@@ -1967,12 +1982,13 @@ Status WriteBatchInternal::InsertInto(
...
     bool ignore_missing_column_families, uint64_t recovery_log_number, DB* db,
-    bool concurrent_memtable_writes, bool seq_per_batch, bool batch_per_txn) {
+    bool concurrent_memtable_writes, bool seq_per_batch, bool batch_per_txn,
+    uint64_t decree, bool pegasus_data) {
   MemTableInserter inserter(
       sequence, memtables, flush_scheduler, trim_history_scheduler,
       ignore_missing_column_families, recovery_log_number, db,
       concurrent_memtable_writes, nullptr /*has_valid_writes*/, seq_per_batch,
-      batch_per_txn);
+      batch_per_txn, false, decree, pegasus_data);
```
Comment (neverchanje) [inline db/write_batch.cc]:
```suggestion
      batch_per_txn, false /*hint_per_batch*/, decree, pegasus_data);
```

File: include/rocksdb/db.h (hunk 7 above)
Comment (acelyc111) [reply include/rocksdb/db.h:1183]: Here code style is keep the same with L1175, format like what?

File: tools/sst_dump_tool.cc (hunk 8 above)
Comment (acelyc111) [reply tools/sst_dump_tool.cc:506]: > In pr 23, If it is not pegasus_data should also print some info. Is it useless?
> code in pr 23:
> 
> ```
>  if (!options_.pegasus_data) {
>         fprintf(stdout, "%s => %s\n",
>             ikey.DebugString(output_hex_).c_str(),
>             value.ToString(output_hex_).c_str());
>       } else {
>         if (ikey.user_key.size() >= 2) {
>           uint32_t expire_ts = 0;
>           std::string hash_key, sort_key, user_data;
>           pegasus_restore_key(ikey.user_key, hash_key, sort_key);
>           pegasus_restore_value(value, expire_ts, user_data);
>           std::ostringstream oss;
>           oss << "\"" << escape_string(hash_key) << "\" : \"" << escape_string(sort_key)
>               << "\" @ " << ikey.sequence << " : " << ikey.type << " => "
>               << expire_ts << " : \"" << escape_string(user_data) << "\"";
>           fprintf(stdout, "%s\n", oss.str().c_str());
>         }
>       }
> ```

Yes, there some code block changes.
Old code:
```
if (not pegasus_data) {
  do A
} else {
  do B
}
```
Now
```
if (pegasus_data) {
  do B
  continue;
}
do A
```
So, there are no logic modifications, just try to keep code difference as small as possible.

File: db/version_edit.cc [hunk 12]
```diff
@@ -695,6 +761,14 @@ std::string VersionEdit::DebugString(bool hex_key) const {
     AppendNumberTo(&r, remaining_entries_);
     r.append(" entries remains");
   }
+  if (has_pegasus_data_version_) {
+    r.append("\n  ValueSchemaVersion: ");
```
Comment (neverchanje) [inline db/version_edit.cc]:
```suggestion
    r.append("\n  PegasusDataVersion: ");
```

File: util/compression.h (hunk 10 above)
Comment (acelyc111) [reply util/compression.h:1051]: Yes, this is a bug of rocksdb, `ctx` is not defined, on some server it will report a compile error (our build server, or Pegasus CI, I'm not sure now). And I've submmited this patch to rocksdb official repo.

File: include/rocksdb/db.h (hunk 7 above)
Comment (levy5307) [reply include/rocksdb/db.h:1183]: In pr 23, these code is like this:
```
  // Retrieve the list of all files in the database.
  virtual Status GetLiveFilesQuick(std::vector<std::string>& ret,
                                   uint64_t* manifest_file_size,
                                   SequenceNumber* last_sequence,
                                   uint64_t* last_decree) const { return Status::NotSupported(); }
```
so I guess these code is not format in this pr

File: db/version_set.cc [hunk 13]
```diff
@@ -4501,6 +4565,14 @@ Status VersionSet::Recover(
                                current_version_number_++);
       builder->SaveTo(v->storage_info());
 
+      if (db_options_->pegasus_data) {
+        // update last flush sequence/decree
+        SequenceNumber sequence;
+        uint64_t decree = 0;
+        version_edit_params.GetLastFlushSeqDecree(&sequence, &decree);
```
Comment (neverchanje) [inline db/version_set.cc:4572]: Previous version
```
        auto &p = last_flush_seq_decree_map[cfd->GetID()];
```
Current version
```
        SequenceNumber sequence;
        uint64_t decree = 0;
        version_edit_params.GetLastFlushSeqDecree(&sequence, &decree);
```
Is this change safe?
```
  void GetLastFlushSeqDecree(SequenceNumber* sequence, uint64_t* decree) const {
    *sequence = last_flush_sequence_;
    *decree = last_flush_decree_;
  }
```

File: include/rocksdb/db.h (hunk 7 above)
Comment (neverchanje) [reply include/rocksdb/db.h:1183]: RocksDB team doesn't emphasize code formatting in their review. So will we do. I think there's no rule to restrict whether the function should be in one line.

File: db/version_set.cc (hunk 13 above)
Comment (acelyc111) [reply db/version_set.cc:4572]: Yes, it's safe. The main difference is that the new version has made much refactor, encapsulation some code to functions, and now there is no such a `last_flush_seq_decree_map`  in `VersionSet::Recover`, version edit reader has been move to `ReadAndRecover`, `ApplyOneVersionEditToBuilder` and `ExtractInfoFromVersionEdit`, and results are put into `VersionEditParams* version_edit_params`, so we use `version_edit_params ` here now.

tips: There is a variable named `last_flush_seq_decree_map`, but it's in `DumpManifest`.

**Review by neverchanje — APPROVED** [review]

**Review by levy5307 — APPROVED** [review]
